File: js/src/twitter/createReverseTwitterRegistry.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createReverseTwitterRegistry = void 0;
const borsh_1 = require("borsh");
const buffer_1 = require("buffer");
const web3_js_1 = require("@solana/web3.js");
const constants_1 = require("../constants");
const createInstruction_1 = require("../instructions/createInstruction");
const updateInstruction_1 = require("../instructions/updateInstruction");
const state_1 = require("../state");
const getHashedNameSync_1 = require("../utils/getHashedNameSync");
const getNameAccountKeySync_1 = require("../utils/getNameAccountKeySync");
const int_1 = require("../int");
const ReverseTwitterRegistryState_1 = require("./ReverseTwitterRegistryState");
function createReverseTwitterRegistry(connection, twitterHandle, twitterRegistryKey, verifiedPubkey, payerKey) {
    return __awaiter(this, void 0, void 0, function* () {
        // Create the reverse lookup registry
        const hashedVerifiedPubkey = (0, getHashedNameSync_1.getHashedNameSync)(verifiedPubkey.toString());
        const reverseRegistryKey = (0, getNameAccountKeySync_1.getNameAccountKeySync)(hashedVerifiedPubkey, constants_1.TWITTER_VERIFICATION_AUTHORITY, constants_1.TWITTER_ROOT_PARENT_REGISTRY_KEY);
        let reverseTwitterRegistryStateBuff = (0, borsh_1.serialize)(ReverseTwitterRegistryState_1.ReverseTwitterRegistryState.schema, new ReverseTwitterRegistryState_1.ReverseTwitterRegistryState({
            twitterRegistryKey: twitterRegistryKey.toBytes(),
            twitterHandle,
        }));
        return [
            (0, createInstruction_1.createInstruction)(constants_1.NAME_PROGRAM_ID, web3_js_1.SystemProgram.programId, reverseRegistryKey, verifiedPubkey, payerKey, hashedVerifiedPubkey, new int_1.Numberu64(yield connection.getMinimumBalanceForRentExemption(reverseTwitterRegistryStateBuff.length + state_1.NameRegistryState.HEADER_LEN)), new int_1.Numberu32(reverseTwitterRegistryStateBuff.length), constants_1.TWITTER_VERIFICATION_AUTHORITY, // Twitter authority acts as class for all reverse-lookup registries
            constants_1.TWITTER_ROOT_PARENT_REGISTRY_KEY, // Reverse registries are also children of the root
            constants_1.TWITTER_VERIFICATION_AUTHORITY),
            (0, updateInstruction_1.updateInstruction)(constants_1.NAME_PROGRAM_ID, reverseRegistryKey, new int_1.Numberu32(0), buffer_1.Buffer.from(reverseTwitterRegistryStateBuff), constants_1.TWITTER_VERIFICATION_AUTHORITY),
        ];
    });
}
exports.createReverseTwitterRegistry = createReverseTwitterRegistry;

File: python/src/twitter/create_reverse_twitter_registry.py
from typing import List

from solana.constants import SYSTEM_PROGRAM_ID
from solana.rpc.async_api import AsyncClient
from solders.instruction import Instruction
from solders.pubkey import Pubkey

from NameRegistryState import NameRegistryState
from constants import TWITTER_VERIFICATION_AUTHORITY, TWITTER_ROOT_PARENT_REGISTRY_KEY, NAME_PROGRAM_ID
from exception import NoLamportsDataException
from instructions import create_instruction
from instructions.update_instruction import update_instruction
from int import Numberu64, Numberu32
from twitter.ReverseTwitterRegistryState import ReverseTwitterRegistryState
from utils import get_hashed_name, get_name_account_key


async def create_reverse_twitter_registry(
        connection: AsyncClient,
        twitter_handle: str,
        twitter_registry: Pubkey,
        verified_pubkey: Pubkey,
        payer: Pubkey,
) -> List[Instruction]:
    hashed_verified_pubkey = get_hashed_name(str(verified_pubkey))
    reverse_registry = get_name_account_key(
        hashed_verified_pubkey,
        TWITTER_VERIFICATION_AUTHORITY,
        TWITTER_ROOT_PARENT_REGISTRY_KEY
    )

    reverse_twitter_registry_state_bytes = ReverseTwitterRegistryState.serialize(
        ReverseTwitterRegistryState(
            handle=twitter_handle,
            registry=bytes(twitter_registry),
        )
    )
    space = len(reverse_twitter_registry_state_bytes)
    lamports = await connection.get_minimum_balance_for_rent_exemption(
        space + NameRegistryState.HEADER_LEN
    )
    if lamports.value is None:
        raise NoLamportsDataException("Failed to get minimum balance for rent exemption")

    ix_create = create_instruction(
        NAME_PROGRAM_ID,
        SYSTEM_PROGRAM_ID,
        reverse_registry,
        verified_pubkey,
        payer,
        hashed_verified_pubkey,
        Numberu64(lamports.value),
        Numberu32(space),
        TWITTER_VERIFICATION_AUTHORITY,
        TWITTER_ROOT_PARENT_REGISTRY_KEY,
        TWITTER_VERIFICATION_AUTHORITY
    )

    ix_update = update_instruction(
        NAME_PROGRAM_ID,
        reverse_registry,
        Numberu32(0),
        reverse_twitter_registry_state_bytes,
        TWITTER_VERIFICATION_AUTHORITY
    )

    return [ix_create, ix_update]